File: test/update.test.ts
import { afterEach, describe, expect, test } from "bun:test";
import { chmod, mkdtemp, rm, writeFile } from "node:fs/promises";
import { tmpdir } from "node:os";
import { join } from "node:path";
import {
  compareSemver,
  fetchLatestPublishedRelease,
  leavePluginRoot,
  parsePluginListSource,
  parseReleaseTag,
  ReleaseCheckError,
  runUpdate,
} from "../src/update";
import { PRODUCT_VERSION } from "../src/version";

const dirs: string[] = [];
afterEach(async () => {
  await Promise.all(dirs.splice(0).map((d) => rm(d, { recursive: true, force: true })));
});

const NEWER = "0.999.0";

async function writeFakeHerdr(pluginListJson: string): Promise<string> {
  const dir = await mkdtemp(join(tmpdir(), "hwf-update-herdr-"));
  dirs.push(dir);
  const bin = join(dir, "fake-herdr");
  const script = `#!/bin/sh
if [ "$1" = "plugin" ] && [ "$2" = "list" ]; then
  printf '%s\\n' ${JSON.stringify(pluginListJson)}
  exit 0
fi
exit 1
`;
  await writeFile(bin, script);
  await chmod(bin, 0o755);
  return bin;
}

function githubListJson(): string {
  return JSON.stringify({
    id: "cli:plugin",
    result: {
      type: "plugin_list",
      plugins: [
        {
          plugin_id: "herdr-workflows",
          source: { kind: "github", owner: "aorumbayev", repo: "herdr-workflows" },
        },
      ],
    },
  });
}

function localListJson(): string {
  return JSON.stringify({
    id: "cli:plugin",
    result: {
      type: "plugin_list",
      plugins: [{ plugin_id: "herdr-workflows", source: { kind: "local" } }],
    },
  });
}

function emptyListJson(): string {
  return JSON.stringify({
    id: "cli:plugin",
    result: { type: "plugin_list", plugins: [] },
  });
}

async function captureRunUpdate(
  deps: Parameters<typeof runUpdate>[0],
  env: Record<string, string | undefined> = {},
): Promise<{ stdout: string; stderr: string; exitCode?: number; error?: unknown }> {
  const stdout: string[] = [];
  const stderr: string[] = [];
  const origOut = process.stdout.write.bind(process.stdout);
  const origErr = process.stderr.write.bind(process.stderr);
  const origExit = process.exit;
  const saved: Record<string, string | undefined> = {};
  for (const key of Object.keys(env)) {
    saved[key] = process.env[key];
    const value = env[key];
    if (value === undefined) delete process.env[key];
    else process.env[key] = value;
  }
  let exitCode: number | undefined;
  process.stdout.write = ((chunk: string | Uint8Array) => {
    stdout.push(typeof chunk === "string" ? chunk : Buffer.from(chunk).toString());
    return true;
  }) as typeof process.stdout.write;
  process.stderr.write = ((chunk: string | Uint8Array) => {
    stderr.push(typeof chunk === "string" ? chunk : Buffer.from(chunk).toString());
    return true;
  }) as typeof process.stderr.write;
  process.exit = ((code?: number) => {
    exitCode = code ?? 0;
    throw new Error(`exit:${exitCode}`);
  }) as typeof process.exit;
  const beforeCwd = process.cwd();
  try {
    await runUpdate(deps);
    return { stdout: stdout.join(""), stderr: stderr.join(""), exitCode };
  } catch (error) {
    if (error instanceof Error && error.message.startsWith("exit:")) {
      return { stdout: stdout.join(""), stderr: stderr.join(""), exitCode };
    }
    return { stdout: stdout.join(""), stderr: stderr.join(""), exitCode, error };
  } finally {
    process.stdout.write = origOut;
    process.stderr.write = origErr;
    process.exit = origExit;
    for (const [key, value] of Object.entries(saved)) {
      if (value === undefined) delete process.env[key];
      else process.env[key] = value;
    }
    process.chdir(beforeCwd);
  }
}

describe("release-check", () => {
  test("parses strict v0.x.y tags and rejects others", () => {
    expect(parseReleaseTag("v0.2.3")).toEqual({ tag: "v0.2.3", version: "0.2.3" });
    expect(() => parseReleaseTag("0.2.3")).toThrow(ReleaseCheckError);
    expect(() => parseReleaseTag("v1.0.0")).toThrow(ReleaseCheckError);
    expect(() => parseReleaseTag("v0.2.3-beta")).toThrow(ReleaseCheckError);
  });

  test("compares 0.x.y versions", () => {
    expect(compareSemver("0.1.0", "0.2.0")).toBeLessThan(0);
    expect(compareSemver("0.2.0", "0.2.0")).toBe(0);
    expect(compareSemver("0.3.1", "0.2.9")).toBeGreaterThan(0);
  });

  test("fetchLatestPublishedRelease validates JSON and respects timeout", async () => {
    const ok = await fetchLatestPublishedRelease({
      fetchImpl: async () =>
        new Response(JSON.stringify({ tag_name: "v0.4.0", draft: false }), { status: 200 }),
    });
    expect(ok).toEqual({ tag: "v0.4.0", version: "0.4.0" });

    await expect(
      fetchLatestPublishedRelease({
        fetchImpl: async () => new Response("nope", { status: 404 }),
      }),
    ).rejects.toThrow(/HTTP 404/);

    await expect(
      fetchLatestPublishedRelease({
        fetchImpl: async () =>
          new Response(JSON.stringify({ tag_name: "v0.4.0", draft: true }), { status: 200 }),
      }),
    ).rejects.toThrow(/draft/);

    await expect(
      fetchLatestPublishedRelease({
        timeoutMs: 20,
        fetchImpl: async (_url, init) => {
          await new Promise<void>((resolve, reject) => {
            const t = setTimeout(resolve, 500);
            init?.signal?.addEventListener("abort", () => {
              clearTimeout(t);
              reject(Object.assign(new Error("aborted"), { name: "AbortError" }));
            });
          });
          return new Response("{}");
        },
      }),
    ).rejects.toThrow(/timed out/);
  });
});

describe("update", () => {
  test("no-ops when current or newer", async () => {
    const result = await captureRunUpdate({
      fetchLatest: async () => ({ tag: `v${PRODUCT_VERSION}`, version: PRODUCT_VERSION }),
    });
    expect(result.exitCode).toBeUndefined();
    expect(result.error).toBeUndefined();
    expect(result.stdout).toContain(`already up to date (${PRODUCT_VERSION})`);
  });

  test("refuses linked development checkouts", async () => {
    const herdr = await writeFakeHerdr(localListJson());
    const result = await captureRunUpdate(
      { fetchLatest: async () => ({ tag: `v${NEWER}`, version: NEWER }) },
      { HERDR_BIN_PATH: herdr },
    );
    expect(result.exitCode).toBe(1);
    expect(result.stderr).toMatch(/bun run install:dev/);
  });

  test("explains unregistered binaries", async () => {
    const herdr = await writeFakeHerdr(emptyListJson());
    const result = await captureRunUpdate(
      { fetchLatest: async () => ({ tag: `v${NEWER}`, version: NEWER }) },
      { HERDR_BIN_PATH: herdr },
    );
    expect(result.exitCode).toBe(1);
    expect(result.stderr).toMatch(/herdr plugin install aorumbayev\/herdr-workflows/);
  });

  test("leaves plugin root and forwards install failure", async () => {
    const herdr = await writeFakeHerdr(githubListJson());
    const pluginRoot = await mkdtemp(join(tmpdir(), "hwf-managed-plugin-"));
    dirs.push(pluginRoot);
    const installs: Array<{ args: string[]; cwd: string }> = [];
    const result = await captureRunUpdate(
      {
        fetchLatest: async () => ({ tag: `v${NEWER}`, version: NEWER }),
        runInstall: async (args, cwd) => {
          installs.push({ args, cwd });
          return 7;
        },
      },
      { HERDR_BIN_PATH: herdr, HERDR_PLUGIN_ROOT: pluginRoot },
    );
    expect(result.exitCode).toBe(7);
    expect(installs).toHaveLength(1);
    expect(installs[0]!.args).toEqual(["plugin", "install", "aorumbayev/herdr-workflows", "--yes"]);
    expect(installs[0]!.cwd).not.toBe(pluginRoot);
    expect(result.stdout).toContain(`updating ${PRODUCT_VERSION} → ${NEWER}`);
    expect(result.stderr).toContain("exit 7");
  });

  test("successful managed install", async () => {
    const herdr = await writeFakeHerdr(githubListJson());
    const pluginRoot = await mkdtemp(join(tmpdir(), "hwf-managed-plugin-"));
    dirs.push(pluginRoot);
    const result = await captureRunUpdate(
      {
        fetchLatest: async () => ({ tag: `v${NEWER}`, version: NEWER }),
        runInstall: async () => 0,
      },
      { HERDR_BIN_PATH: herdr, HERDR_PLUGIN_ROOT: pluginRoot },
    );
    expect(result.exitCode).toBeUndefined();
    expect(result.error).toBeUndefined();
    expect(result.stdout).toContain(`updated to ${NEWER}`);
  });

  test("parsePluginListSource distinguishes github, local, and missing", () => {
    expect(parsePluginListSource(githubListJson())).toEqual({
      kind: "github",
      owner: "aorumbayev",
      repo: "herdr-workflows",
    });
    expect(parsePluginListSource(localListJson())).toEqual({ kind: "local" });
    expect(parsePluginListSource(emptyListJson())).toEqual({ kind: "unregistered" });
    expect(
      parsePluginListSource(
        JSON.stringify({
          result: {
            type: "plugin_list",
            plugins: [{ plugin_id: "other", source: { kind: "github" } }],
          },
        }),
      ),
    ).toEqual({ kind: "unregistered" });
    expect(
      parsePluginListSource(JSON.stringify({ plugins: [{ plugin_id: "herdr-workflows" }] })),
    ).toEqual({
      kind: "unregistered",
    });
  });

  test("leavePluginRoot moves cwd outside the managed checkout", () => {
    const before = process.cwd();
    try {
      const outside = leavePluginRoot(before, {
        ...process.env,
        HOME: process.env.HOME || before,
      });
      expect(outside).not.toBe(before);
    } finally {
      process.chdir(before);
    }
  });

  test("update help is registered without loading the picker", async () => {
    const proc = Bun.spawn([process.execPath, "src/cli.ts", "update", "--help"], {
      cwd: join(import.meta.dir, ".."),
      stdout: "pipe",
      stderr: "pipe",
      env: {
        ...process.env,
        HERDR_BIN_PATH: "/does-not-exist/herdr",
        HERDR_SOCKET_PATH: "",
      },
    });
    const [stdout, code] = await Promise.all([new Response(proc.stdout).text(), proc.exited]);
    expect(code).toBe(0);
    expect(stdout).toMatch(/Update to the latest published GitHub Release/);
  });
});

File: internal/update/update_test.go
package update

import (
	"encoding/json"
	"errors"
	"net/http"
	"net/http/httptest"
	"os"
	"path/filepath"
	"testing"
	"time"
)

func githubListJSON() string {
	return `{"id":"cli:plugin","result":{"type":"plugin_list","plugins":[{"plugin_id":"herdr-workflows","source":{"kind":"github","owner":"aorumbayev","repo":"herdr-workflows"}}]}}`
}

func localListJSON() string {
	return `{"id":"cli:plugin","result":{"type":"plugin_list","plugins":[{"plugin_id":"herdr-workflows","source":{"kind":"local"}}]}}`
}

func emptyListJSON() string {
	return `{"id":"cli:plugin","result":{"type":"plugin_list","plugins":[]}}`
}

func TestParseReleaseTagAndCompare(t *testing.T) {
	got, err := ParseReleaseTag("v0.2.3")
	if err != nil || got.Tag != "v0.2.3" || got.Version != "0.2.3" {
		t.Fatalf("%+v %v", got, err)
	}
	if _, err := ParseReleaseTag("0.2.3"); err == nil {
		t.Fatal("bare tag")
	}
	if _, err := ParseReleaseTag("v1.0.0"); err == nil {
		t.Fatal("v1")
	}
	if _, err := ParseReleaseTag("v0.2.3-beta"); err == nil {
		t.Fatal("prerelease")
	}
	if n, _ := CompareSemver("0.1.0", "0.2.0"); n >= 0 {
		t.Fatal("0.1 < 0.2")
	}
	if n, _ := CompareSemver("0.2.0", "0.2.0"); n != 0 {
		t.Fatal("equal")
	}
	if n, _ := CompareSemver("0.3.1", "0.2.9"); n <= 0 {
		t.Fatal("0.3.1 > 0.2.9")
	}
}

func TestCheckForUpdateValidatesJSONAndTimeout(t *testing.T) {
	okSrv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		_ = json.NewEncoder(w).Encode(map[string]any{"tag_name": "v0.4.0", "draft": false})
	}))
	defer okSrv.Close()
	got, err := CheckForUpdate(CheckOpts{URL: okSrv.URL, Client: okSrv.Client()})
	if err != nil || got.Version != "0.4.0" {
		t.Fatalf("%+v %v", got, err)
	}
	bad := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(404)
		_, _ = w.Write([]byte("nope"))
	}))
	defer bad.Close()
	_, err = CheckForUpdate(CheckOpts{URL: bad.URL, Client: bad.Client()})
	var checkErr *ReleaseCheckError
	if err == nil || !errors.As(err, &checkErr) || checkErr.Error() != "latest release request failed: HTTP 404" {
		t.Fatalf("404 err %v", err)
	}
	draft := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		_ = json.NewEncoder(w).Encode(map[string]any{"tag_name": "v0.4.0", "draft": true})
	}))
	defer draft.Close()
	if _, err := CheckForUpdate(CheckOpts{URL: draft.URL, Client: draft.Client()}); err == nil || err.Error() != "latest release endpoint returned a draft" {
		t.Fatalf("draft %v", err)
	}
	hang := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		time.Sleep(200 * time.Millisecond)
	}))
	defer hang.Close()
	if _, err := CheckForUpdate(CheckOpts{URL: hang.URL, Client: hang.Client(), Timeout: 20 * time.Millisecond}); err == nil {
		t.Fatal("expected timeout")
	} else if _, ok := err.(*ReleaseCheckError); !ok {
		t.Fatalf("timeout type %T %v", err, err)
	}
}

func TestParsePluginListSource(t *testing.T) {
	gh, err := ParsePluginListSource(githubListJSON())
	if err != nil || gh.Kind != "github" || gh.Owner != "aorumbayev" || gh.Repo != "herdr-workflows" {
		t.Fatalf("%+v %v", gh, err)
	}
	local, _ := ParsePluginListSource(localListJSON())
	if local.Kind != "local" {
		t.Fatalf("%+v", local)
	}
	empty, _ := ParsePluginListSource(emptyListJSON())
	if empty.Kind != "unregistered" {
		t.Fatalf("%+v", empty)
	}
	other, _ := ParsePluginListSource(`{"result":{"type":"plugin_list","plugins":[{"plugin_id":"other","source":{"kind":"github"}}]}}`)
	if other.Kind != "unregistered" {
		t.Fatalf("%+v", other)
	}
	bare, _ := ParsePluginListSource(`{"plugins":[{"plugin_id":"herdr-workflows"}]}`)
	if bare.Kind != "unregistered" {
		t.Fatalf("%+v", bare)
	}
}

func TestLeavePluginRootOutsideCheckout(t *testing.T) {
	before, err := os.Getwd()
	if err != nil {
		t.Fatal(err)
	}
	outside, err := LeavePluginRoot(before, os.Getenv)
	if err != nil {
		t.Fatal(err)
	}
	if outside == before {
		t.Fatal("did not leave")
	}
	wd, _ := os.Getwd()
	if wd != before {
		t.Fatalf("process cwd mutated to %s", wd)
	}
}

func TestUpdatePluginOutcomes(t *testing.T) {
	current := "0.1.0"
	newer := "0.999.0"
	up, err := UpdatePlugin(Deps{
		Version:     current,
		FetchLatest: func() (LatestRelease, error) { return LatestRelease{Tag: "v" + current, Version: current}, nil },
	})
	if err != nil || up.Kind != "up_to_date" || up.Current != current {
		t.Fatalf("%+v %v", up, err)
	}
	local, err := UpdatePlugin(Deps{
		Version:     current,
		FetchLatest: func() (LatestRelease, error) { return LatestRelease{Tag: "v" + newer, Version: newer}, nil },
		ListSource:  func() (PluginSourceInfo, error) { src, _ := ParsePluginListSource(localListJSON()); return src, nil },
	})
	if err != nil || local.Kind != "refused_local" {
		t.Fatalf("%+v %v", local, err)
	}
	standaloneDir := t.TempDir()
	standaloneDest := filepath.Join(standaloneDir, "herdr-workflows")
	if err := os.WriteFile(standaloneDest, []byte("old"), 0o755); err != nil {
		t.Fatal(err)
	}
	var standaloneCalls int
	unreg, err := UpdatePlugin(Deps{
		Version:     current,
		FetchLatest: func() (LatestRelease, error) { return LatestRelease{Tag: "v" + newer, Version: newer}, nil },
		ListSource:  func() (PluginSourceInfo, error) { src, _ := ParsePluginListSource(emptyListJSON()); return src, nil },
		Executable:  func() (string, error) { return standaloneDest, nil },
		InstallRelease: func(opts InstallOpts) error {
			standaloneCalls++
			if opts.Version != newer || opts.DestPath != standaloneDest {
				t.Fatalf("opts %+v", opts)
			}
			return os.WriteFile(opts.DestPath, []byte("new"), 0o755)
		},
	})
	if err != nil || unreg.Kind != "updated" || unreg.To != newer {
		t.Fatalf("%+v %v", unreg, err)
	}
	if standaloneCalls != 1 {
		t.Fatalf("standaloneCalls = %d", standaloneCalls)
	}
	pluginRoot := t.TempDir()
	var installs []struct {
		args []string
		cwd  string
	}
	fail, err := UpdatePlugin(Deps{
		Version:     current,
		PluginRoot:  pluginRoot,
		FetchLatest: func() (LatestRelease, error) { return LatestRelease{Tag: "v" + newer, Version: newer}, nil },
		ListSource:  func() (PluginSourceInfo, error) { src, _ := ParsePluginListSource(githubListJSON()); return src, nil },
		RunInstall: func(args []string, cwd string) (int, error) {
			installs = append(installs, struct {
				args []string
				cwd  string
			}{args, cwd})
			return 7, nil
		},
		Getenv: os.Getenv,
	})
	if err != nil || fail.Kind != "install_failed" || fail.Code != 7 {
		t.Fatalf("%+v %v", fail, err)
	}
	if len(installs) != 1 || installs[0].cwd == pluginRoot {
		t.Fatalf("installs %+v", installs)
	}
	wantArgs := []string{"plugin", "install", "aorumbayev/herdr-workflows", "--ref", "v" + newer, "--yes"}
	if len(installs[0].args) != len(wantArgs) {
		t.Fatalf("args %v", installs[0].args)
	}
	for i := range wantArgs {
		if installs[0].args[i] != wantArgs[i] {
			t.Fatalf("args %v", installs[0].args)
		}
	}
	ok, err := UpdatePlugin(Deps{
		Version:     current,
		PluginRoot:  pluginRoot,
		FetchLatest: func() (LatestRelease, error) { return LatestRelease{Tag: "v" + newer, Version: newer}, nil },
		ListSource:  func() (PluginSourceInfo, error) { src, _ := ParsePluginListSource(githubListJSON()); return src, nil },
		RunInstall:  func(args []string, cwd string) (int, error) { return 0, nil },
		Getenv:      os.Getenv,
	})
	if err != nil || ok.Kind != "updated" || ok.To != newer {
		t.Fatalf("%+v %v", ok, err)
	}
}
